deploy: update csi-provisioner to latest version

yati1998 opened the pull request:
This commits updates csi-provisioner sidecar to
latest version i.e., v3.2.0.

Refer this for more details:
https://github.com/kubernetes-csi/external-provisioner/releases/tag/v3.2.0

Signed-off-by: Yati Padia <ypadia@redhat.com>

yati1998 committed:
deploy: update csi-provisioner to latest version

This commits updates csi-provisioner sidecar to
latest version i.e., v3.2.0.

fixes: #3184

Signed-off-by: Yati Padia <ypadia@redhat.com>

Madhu-1 commented:
>Error: initializing source docker://registry.k8s.io/sig-storage/csi-provisioner:v3.2.0: reading manifest v3.2.0 in registry.k8s.io/sig-storage/csi-provisioner: manifest unknown: Failed to fetch "v3.2.0" from request "/v2/sig-storage/csi-provisioner/manifests/v3.2.0".


Looks like the image is not available yet.

Madhu-1 commented on deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:
```diff
@@ -43,14 +43,15 @@ spec:
             - "--leader-election=true"
             - "--retry-interval-start=500ms"
             - "--feature-gates=Topology=false"
+            - "--feature-gates=HonorPVReclaimPolicy=true"
```
the same change is required in helm deployment yaml and for rbd also

yati1998 replied:
okay

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-deployment.yaml:
```diff
@@ -62,6 +62,7 @@ spec:
             - "--leader-election=true"
             - "--retry-interval-start=500ms"
             - "--extra-create-metadata=true"
+            - "--feature-gates=HonorPVReclaimPolicy=true"
```
what about `prevent-volume-mode-conversion` which blocks the change of volumeMode? Do we need to support it?

cc @humblec @nixpanic

humblec replied:
Yeah, we have to enable the prevent-volume-mode-conversion ( `= true`) and keep it safe.  @Madhu-1 @yati1998

yati1998 replied:
cool

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-deployment.yaml:
```diff
@@ -62,6 +62,8 @@ spec:
             - "--retry-interval-start=500ms"
             - "--extra-create-metadata=true"
+            - "--feature-gates=HonorPVReclaimPolicy=true"
+            - "--feature-gates=prevent-volume-mode-conversion=true"
```
this is not a feature-gate its just a argument

humblec commented on deploy/rbd/kubernetes/csi-rbdplugin-provisioner.yaml:
```diff
@@ -60,6 +60,8 @@ spec:
             #  set it to true to use topology based provisioning
             - "--feature-gates=Topology=false"
+            - "--feature-gates=HonorPVReclaimPolicy=true"
+            - "--prevent-volume-mode-conversion=true"
```
@yati1998 I would move this line to other ARGS

humblec replied:
as discussed it looks its not grouped anywhere . so should be ok for now in this format.

humblec reviewed (changes requested):
Please group the flags together , so that we wont miss it.

yati1998 commented:
/retest all

yati1998 commented:
/retest all

yati1998 commented:
/retest all

yati1998 commented:
``` 
GRPC error: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory

E0620 12:51:39.443149       1 controller.go:1481] delete "pvc-4a58b909-129e-4648-a3ca-28df2d6858f3": volume deletion failed: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory

W0620 12:51:39.443174       1 controller.go:989] Retrying syncing volume "pvc-4a58b909-129e-4648-a3ca-28df2d6858f3", failure 7

E0620 12:51:39.443203       1 controller.go:1007] error syncing volume "pvc-4a58b909-129e-4648-a3ca-28df2d6858f3": rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory

I0620 12:51:39.443618       1 event.go:285] Event(v1.ObjectReference{Kind:"PersistentVolume", Namespace:"", Name:"pvc-4a58b909-129e-4648-a3ca-28df2d6858f3", UID:"7bf33e1b-4e5e-4b91-acac-1e50e9d2d78f", APIVersion:"v1", ResourceVersion:"23913", FieldPath:""}): type: 'Warning' reason: 'VolumeFailedDelete' rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
```

[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.21/detail/mini-e2e_k8s-1.21/4148/pipeline)
@nixpanic I see it fails to delete the export, Can you check once?

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.23

yati1998 commented:
/retest ci/centos/upgrade-tests-cephfs

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.23

yati1998 commented:
/retest ci/centos/upgrade-tests-cephfs

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.23

yati1998 commented:
/retest ci/centos/upgrade-tests-cephfs

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.23

yati1998 commented:
/retest ci/centos/upgrade-tests-cephfs

Madhu-1 commented:
@yati1998 can you please check the test failures? is it due to known issues?

yati1998 commented:
`event for csi-nfsplugin-provisioner-86566d4b-89vc7: {kubelet minikube} FailedMount: MountVolume.SetUp failed for volume "ceph-csi-config" : configmap "ceph-csi-config" not found`  [logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3529/pipeline#step-92-log-870)

yati1998 commented:
It fails to delete the volume.
`I0701 06:06:29.087674       1 connection.go:183] GRPC call: /csi.v1.Controller/DeleteVolume
I0701 06:06:29.087704       1 connection.go:184] GRPC request: {"secrets":"***stripped***","volume_id":"0001-0024-49561584-49f7-4c72-96e8-67e3611f9e13-0000000000000001-ed0f85e4-f903-11ec-a1a6-6ee75cf27672"}
I0701 06:06:29.113302       1 connection.go:186] GRPC response: {}
I0701 06:06:29.113369       1 connection.go:187] GRPC error: <nil>
I0701 06:06:29.113386       1 controller.go:1486] delete "pvc-1acd6d4a-67d0-48da-b942-757f7970258a": volume deleted
I0701 06:06:29.120800       1 controller.go:1523] delete "pvc-1acd6d4a-67d0-48da-b942-757f7970258a": failed to remove finalizer for persistentvolume: persistentvolumes "pvc-1acd6d4a-67d0-48da-b942-757f7970258a" is forbidden: User "system:serviceaccount:cephcsi-e2e-8301cf6b:cephfs-csi-provisioner" cannot update resource "persistentvolumes" in API group "" at the cluster scope
W0701 06:06:29.120868       1 controller.go:989] Retrying syncing volume "pvc-1acd6d4a-67d0-48da-b942-757f7970258a", failure 1
E0701 06:06:29.120914       1 controller.go:1007] error syncing volume "pvc-1acd6d4a-67d0-48da-b942-757f7970258a": persistentvolumes "pvc-1acd6d4a-67d0-48da-b942-757f7970258a" is forbidden: User "system:serviceaccount:cephcsi-e2e-8301cf6b:cephfs-csi-provisioner" cannot update resource "persistentvolumes" in API group "" at the cluster scope`
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.23/runs/2161/nodes/89/steps/92/log/?start=0)

yati1998 commented:
```I0704 14:12:07.638723       1 connection.go:183] GRPC call: /csi.v1.Controller/DeleteVolume
I0704 14:12:07.638748       1 connection.go:184] GRPC request: {"secrets":"***stripped***","volume_id":"0001-0024-5b6288f7-6ecc-418f-87cd-4f95212614a2-0000000000000001-0c9124dd-fba3-11ec-876a-36728d15ed92"}
I0704 14:12:07.644920       1 connection.go:186] GRPC response: {}
I0704 14:12:07.644969       1 connection.go:187] GRPC error: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
E0704 14:12:07.645012       1 controller.go:1481] delete "pvc-7522b7f4-5b6f-4ded-af22-29c65d279d46": volume deletion failed: rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
W0704 14:12:07.645097       1 controller.go:989] Retrying syncing volume "pvc-7522b7f4-5b6f-4ded-af22-29c65d279d46", failure 0
E0704 14:12:07.645161       1 controller.go:1007] error syncing volume "pvc-7522b7f4-5b6f-4ded-af22-29c65d279d46": rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory
I0704 14:12:07.645211       1 event.go:285] Event(v1.ObjectReference{Kind:"PersistentVolume", Namespace:"", Name:"pvc-7522b7f4-5b6f-4ded-af22-29c65d279d46", UID:"69abed12-1a32-4d94-b00e-6736191c89bf", APIVersion:"v1", ResourceVersion:"2059", FieldPath:""}): type: 'Warning' reason: 'VolumeFailedDelete' rpc error: code = InvalidArgument desc = failed to delete export: failed to identify NFS cluster: failed to get cluster name: failed to get values for key "csi.nfs.cluster" from OMAP: key not found: rados: ret=-2, No such file or directory```

Rakshith-R commented:
Adding DNM label, to debug the issue first. 

@yati1998 
Similar rbac changes are required for nfs too ?
https://github.com/ceph/ceph-csi/pull/3186/files#diff-1c80efeb487f6d4fd9fc8c16017a83a4b25953066c5146057adf9a756ca6714b

Rakshith-R reviewed (changes requested):
@yati1998 , 
nfs pvc not found error handling needs to be implemented during create volume stage for pvhonorreclaim thing, 
I vaguely remember you were gonna work on it, right ?  
https://github.com/ceph/ceph-csi/issues/2913
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2644-honor-pv-reclaim-policy

cc @nixpanic @Madhu-1 

We can open another issue for that and remove it as part of this pr.

Madhu-1 commented:
> @yati1998 , nfs pvc not found error handling needs to be implemented during create volume stage for pvhonorreclaim thing, I vaguely remember you were gonna work on it, right ? #2913 https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2644-honor-pv-reclaim-policy
> 
> cc @nixpanic @Madhu-1
> 
> We can open another issue for that and remove it as part of this pr.

Should be fine or fix the NFS part and rebase this PR to make sure it's fixed(this PR can wait for it no hurry as we don't have any release planned yet). or get this in first without a new feature-gate.

yati1998 commented:
> @yati1998 , nfs pvc not found error handling needs to be implemented during create volume stage for pvhonorreclaim thing, I vaguely remember you were gonna work on it, right ? #2913 https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2644-honor-pv-reclaim-policy

Yeah, I am on it, will check that.

yati1998 committed:
cephfs: add update rbac rule to pv resource

This commit adds the update rbac rule to persistent
volume resource as the ci was failing with below error:
cannot update resource "persistentvolumes" in API group
"" at the cluster scope

Signed-off-by: Yati Padia <ypadia@redhat.com>

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

Rakshith-R commented on charts/ceph-csi-cephfs/templates/provisioner-deployment.yaml:
```diff
@@ -63,6 +63,9 @@ spec:
             - "--leader-election=true"
             - "--retry-interval-start=500ms"
             - "--extra-create-metadata=true"
+            # TODO: enable the feature-gate once error has been handled in nfs
```
Add issue number

Rakshith-R commented on charts/ceph-csi-rbd/templates/provisioner-deployment.yaml:
```diff
@@ -64,6 +64,9 @@ spec:
             - "--retry-interval-start=500ms"
             - "--default-fstype={{ .Values.provisioner.defaultFSType }}"
             - "--extra-create-metadata=true"
+            # TODO: enable the feature-gate once error has been handled in nfs
```
same here

Rakshith-R commented on deploy/rbd/kubernetes/csi-rbdplugin-provisioner.yaml:
```diff
@@ -60,6 +60,9 @@ spec:
             - "--leader-election=true"
```
Change external provisioner version in this deployment too

Rakshith-R reviewed (changes requested):
use v3.2.1 https://github.com/kubernetes-csi/external-provisioner/releases/tag/v3.2.1

Rakshith-R commented on charts/ceph-csi-rbd/templates/provisioner-deployment.yaml:
```diff
@@ -64,6 +64,10 @@ spec:
             - "--retry-interval-start=500ms"
             - "--default-fstype={{ .Values.provisioner.defaultFSType }}"
             - "--extra-create-metadata=true"
+            # TODO: enable the feature-gate once error has been handled in nfs.
```
you can enable the featuregate in rbd and cephfs, disable only in nfs

nixpanic replied:
that's done!

Rakshith-R commented on deploy/cephfs/kubernetes/csi-provisioner-rbac.yaml:
```diff
@@ -18,7 +18,7 @@ rules:
   - apiGroups: [""]
     resources: ["persistentvolumes"]
-    verbs: ["get", "list", "watch", "create", "delete", "patch"]
+    verbs: ["get", "list", "watch", "create", "update", "delete", "patch"]
```
Are you sure this is needed ? 
rbac from external-provisioner does not need it.
https://github.com/kubernetes-csi/external-provisioner/blob/b67655857d4800483928fe62e17660175bcb6ef0/deploy/kubernetes/rbac.yaml#L30

yati1998 replied:
yes, the ci was failing earlier with this error. ```failed to remove finalizer for persistentvolume: persistentvolumes "pvc-1acd6d4a-67d0-48da-b942-757f7970258a" is forbidden: User "system:serviceaccount:cephcsi-e2e-8301cf6b:cephfs-csi-provisioner" cannot update resource "persistentvolumes" in API group "" at the cluster scope```

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

Madhu-1 commented:
@yati1998 still tests are failing?

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/mini-e2e/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

yati1998 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

yati1998 commented:
@Madhu-1 @Rakshith-R CI test seems to be passing now, please add your reviews.

yati1998 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-clusterrole.yaml:
```diff
@@ -15,7 +15,7 @@ rules:
   - apiGroups: [""]
     resources: ["persistentvolumes"]
-    verbs: ["get", "list", "watch", "create", "delete","patch"]
+    verbs: ["get", "list", "watch", "create", "update", "delete","patch"]
```
same RBAC is not required for rbd? :thinking:

yati1998 replied:
rbd has this RBAC already.
